src/language/pattern-types.md

```diff
@@ -0,0 +1,333 @@
+---
+title: Pattern types
+description: Pattern type reference in Dart.
+---
+
+## Cast
+
+`foo as String`
+
+Cast patterns let you insert a [type cast][] as a pattern is
+[destructuring][destructure] values:
+
+```dart
+(num, Object) record = (1, "s");
+var (i as int, s as String) = record;
+```
+
+Cast patterns will [throw][] if the value doesn't have the stated type.
+Like the [null-assert pattern](#null-assert), this lets you forcibly assert the
+expected type of some destructured value.
+
+## Constant	
+
+`123, null, 'string', math.pi, SomeClass.constant, const Thing(1, 2), const (1 + 2)`
+
+Constant patterns allow simple literals and references to named constants to be
+used directly as patterns. They match when the value is equal to the constant.
+
+Some valid constant expressions may not be valid constant patterns. In that case,
+to distinguish a constant expression from a constant pattern, prefix the expression
+with `const` and wrap it in parenthesis so it is unambiguosly an expression and
+not a constant pattern:
+
+```dart
+// List or map pattern:
+case [a, b]: ...
+
+// List or map literal:
+case const [a, b]: ...
+```
+
+## Identifier	
+
+`foo, _`
+
+Identifier patterns may behave like a [constant pattern](#constant) or like a
+[variable pattern](#variable), depending on the context where it appears:
+
+- [Declaration][] context: declares a new variable with identifier name:
+  `var (a, b) = (1, 2);`
+- [Assignment][] context: assigns to existing variable with identifier name:
+  `(a, b) = (3, 4);`
+- [Matching][] context: treated as a named constant pattern (unless its name is `_`):
+  ```dart
+  const c = 1;
+  switch (2) {
+    case c: print('match $c');
+    default: print('no match');    // Prints "no match".
+   }
+  ``` 
+- [Wildcard](#wildcard) identifier in any context: matches any value and discards it:
+  `case [_, var y, _]: print('The middle element is $y');`
+
+## List
+
+`[subpattern1, subpattern2]`
+
+List patterns match values that implement [`List`][], and then recursively match
+its subpatterns against the list's elements to destructure them by position.
+
+List patterns require that the pattern match the entire list, but you can use
```

Comment:
Worth mentioning that in a declaration or assignment context, if the number of elements don't match an exception is thrown?

Reply:
> that any pattern in that context is irrefutable and that an exception is thrown when patterns don't match in any irrefutable context, for any type of pattern, if the value doesn't match match.

Sort of?  In an irrefutable context, mostly the value can't fail to match.  That is, mostly if you see `var <pattern> = expression`, you'll get a static error if the static type of expression doesn't guarantee that the value will match pattern.  There are two bug exceptions: if the type of `expression` is `dynamic`, in which case the match might fail, and if `<pattern>` is a list pattern, in which case it can fail if the length isn't correct.  So there's something fundamentally different between the pattern `(a, b)` and `[a, b]`.  In a refutable context, they either match or don't, but in an irrefutable context, the former will always match, but the latter may throw an exception if the destructured value doesn't have exactly two elements.